app/src/main/java/pm/gnosis/heimdall/ui/safe/recover/extension/ReplaceExtensionQrActivity.kt

```diff
@@ -0,0 +1,185 @@
+                        startActivity(ReplaceExtensionRecoveryPhraseActivity.createIntent(this, safeAddress!!, it.extension))
+                    }
+
+                    is ReplaceExtensionQrContract.PairingResult.PairingError -> {
+                        snackbar(coordinator, R.string.error_pairing_devices)
+                        Timber.e(it.error)
+                    }
+                }
+            }
+        })
+
+        extensionLink.apply {
+            this.text = SpannableStringBuilder(Html.fromHtml(getString(R.string.pairing_first_step)))
+            setOnClickListener { shareExternalText(BuildConfig.CHROME_EXTENSION_URL, "Browser Extension URL") }
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+
+        backArrow.setOnClickListener {
+            finish()
+        }
+
+        disposables += bottomPanel.forwardClicks.subscribe {
+            QRCodeScanActivity.startForResult(this)
+        }
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        handleQrCodeActivityResult(requestCode, resultCode, data, onQrCodeResult = { viewModel.pair(it) })
+    }
+
+    private fun onPairingLoading(isLoading: Boolean) {
+        bottomPanel.disabled = isLoading
+        progressBar.visible(isLoading)
+    }
+
+    companion object {
+
+        private const val EXTRA_SAFE_ADDRESS = "extra.string.safe_address"
+
+        fun createIntent(context: Context, safeAddress: Solidity.Address) = Intent(context, ReplaceExtensionQrActivity::class.java).apply {
+            putExtra(EXTRA_SAFE_ADDRESS, safeAddress.asEthereumAddressString())
+        }
+    }
+}
+
+abstract class ReplaceExtensionQrContract : ViewModel() {
+
+    abstract val observableState: LiveData<ViewUpdate>
+
+    abstract fun setup(safeAddress: Solidity.Address)
+
+    /**
+     * @param signingSafe to indicate which key should be used for pairing. @null if a new key should be generated
+     */
+    abstract fun pair(payload: String)
+
+
+    data class ViewUpdate(
+        val isLoading: Boolean,
+        val pairingResult: PairingResult? = null
+    )
+
+    sealed class PairingResult {
+
+        data class PairingError(val error: Exception) : PairingResult()
+
+        data class PairingSuccess(val safe: Solidity.Address, val extension: Solidity.Address) : PairingResult()
+    }
+}
+
+class ReplaceExtensionQrViewModel @Inject constructor(
+    private val pushServiceRepository: PushServiceRepository,
+    private val moshi: Moshi
+) : ReplaceExtensionQrContract() {
+
+    override val observableState: LiveData<ViewUpdate>
+        get() = _state
+    private val _state = MutableLiveData<ViewUpdate>()
+
+    private lateinit var safeAddress: Solidity.Address
+
+    override fun setup(safeAddress: Solidity.Address) {
+
+        this.safeAddress = safeAddress
+    }
+
+    override fun pair(payload: String) {
+
```

Comment:
If we call `pair` again while a pairing call is still in progress, should we cancel the previous one?